File: infrastructure/persistence/prisma/goal-repository.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PrismaGoalRepository = void 0;
const client_1 = require("./client");
class PrismaGoalRepository {
    async create(goal) {
        const created = await client_1.prisma.goal.create({
            data: {
                id: goal.id,
                userId: goal.userId,
                title: goal.title,
                description: goal.description,
                targetBooks: goal.targetBooks,
                currentBooks: goal.currentBooks,
                startDate: goal.startDate,
                endDate: goal.endDate,
                completed: goal.completed,
            },
        });
        return this.mapToGoal(created);
    }
    async findByUserId(userId) {
        const goals = await client_1.prisma.goal.findMany({
            where: { userId },
            orderBy: { startDate: 'desc' },
        });
        return goals.map(this.mapToGoal);
    }
    async findById(id) {
        const goal = await client_1.prisma.goal.findUnique({
            where: { id },
        });
        if (!goal)
            return undefined;
        return this.mapToGoal(goal);
    }
    async update(id, updates) {
        try {
            const updated = await client_1.prisma.goal.update({
                where: { id },
                data: {
                    ...(updates.title !== undefined && { title: updates.title }),
                    ...(updates.description !== undefined && { description: updates.description }),
                    ...(updates.targetBooks !== undefined && { targetBooks: updates.targetBooks }),
                    ...(updates.currentBooks !== undefined && { currentBooks: updates.currentBooks }),
                    ...(updates.startDate !== undefined && { startDate: updates.startDate }),
                    ...(updates.endDate !== undefined && { endDate: updates.endDate }),
                    ...(updates.completed !== undefined && { completed: updates.completed }),
                },
            });
            return this.mapToGoal(updated);
        }
        catch (error) {
            return null;
        }
    }
    async delete(id) {
        try {
            await client_1.prisma.goal.delete({
                where: { id },
            });
            return true;
        }
        catch (error) {
            return false;
        }
    }
    mapToGoal(prismaGoal) {
        return {
            id: prismaGoal.id,
            userId: prismaGoal.userId,
            title: prismaGoal.title,
            description: prismaGoal.description ?? undefined,
            targetBooks: prismaGoal.targetBooks,
            currentBooks: prismaGoal.currentBooks,
            startDate: prismaGoal.startDate,
            endDate: prismaGoal.endDate,
            completed: prismaGoal.completed,
        };
    }
}
exports.PrismaGoalRepository = PrismaGoalRepository;

File: ios/Packages/InfrastructureIOS/Sources/InfrastructureIOS/Repositories/GoalRepository.swift
import CoreDomain
import Foundation

/// Implementation of GoalRepositoryProtocol using REST API
public final class GoalRepository: GoalRepositoryProtocol {
    private let networkClient: NetworkClientProtocol

    public init(networkClient: NetworkClientProtocol) {
        self.networkClient = networkClient
    }

    // MARK: - GoalRepositoryProtocol

    public func findById(_ id: String) async throws -> Goal? {
        // API doesn't have a single goal endpoint, so get all and filter
        let goals = try await findByUserId("")
        return goals.first { $0.id == id }
    }

    public func findByUserId(_ userId: String) async throws -> [Goal] {
        let endpoint = APIEndpoint.getGoals()
        let response: GetGoalsResponse = try await networkClient.request(endpoint)
        return response.goals.map { $0.toDomain() }
    }
    
    public func findByUserId(_ userId: String, offset: Int, limit: Int?) async throws -> [Goal] {
        // For now, get all goals and apply pagination in memory
        // TODO: Update API to support pagination query parameters
        let allGoals = try await findByUserId(userId)
        let startIndex = min(offset, allGoals.count)
        let endIndex = limit.map { min(startIndex + $0, allGoals.count) } ?? allGoals.count
        return Array(allGoals[startIndex..<endIndex])
    }

    public func create(_ goal: Goal) async throws -> Goal {
        let request = CreateGoalRequest(
            title: goal.title,
            description: goal.description,
            targetBooks: goal.targetBooks,
            startDate: goal.startDate,
            endDate: goal.endDate
        )
        let endpoint = try APIEndpoint.createGoal(request)
        let response: CreateGoalResponse = try await networkClient.request(endpoint)
        return response.goal.toDomain()
    }

    public func update(_ id: String, updates: GoalUpdate) async throws -> Goal? {
        // Build request from updates
        let request = UpdateGoalRequest(
            title: updates.title,
            description: updates.description,
            targetBooks: updates.targetBooks,
            startDate: nil,
            endDate: updates.endDate
        )
        let endpoint = try APIEndpoint.updateGoal(id: id, request)
        let response: UpdateGoalResponse = try await networkClient.request(endpoint)
        return response.goal.toDomain()
    }

    public func delete(_ id: String) async throws -> Bool {
        let endpoint = APIEndpoint.deleteGoal(id: id)
        let _: DeleteGoalResponse = try await networkClient.request(endpoint)
        return true
    }

    public func findActiveByUserId(_ userId: String) async throws -> [Goal] {
        let allGoals = try await findByUserId(userId)
        return allGoals.filter { !$0.completed && $0.endDate >= Date() }
    }
    
    public func exists(userId: String, goalId: String) async throws -> Bool {
        let userGoals = try await findByUserId(userId)
        return userGoals.contains { $0.id == goalId }
    }
}
